# 오류

- 코드의 오류는 불가피함
- 오류 사례를 신중하게 생각하지 않고는 견고하고 신뢰성 높은 코드를 작성할 수 없음

-------

## 복구 가능성

- 오류가 발생했을 때 무엇을 할 것인지 결정하기 위해 자신의 코드가 어떻게 사용될지 신중하게 생각해야 함

### 복구 가능한 오류

- 많은 소프트웨어 오류는 치명적이지 않고, 발생하더라도 사용자는 알아채지 못하도록 적절하게 처리해 작동을 계속할 수 있는 합리적인
방법이 있음
  - 사용자가 잘못된 입력(잘못된 전화번호)를 제공하는 경우 시스템이 작동을 멈추는 것이 아닌 오류 메시지를 제공하고
  올바른 번호를 입력하도록 요청
  - 네트워크 오류(의존하는 서비스)에 연결할 수 없는 경우 몇 초 동안 기다렸다가 다시 시도 혹은 네트워크 연결 확인 요청
  - 중요하지 않은 작업 오류(통계 기록)
- 낮은 층위의 코드에서 오류를 시도하고 복구하는 것은 장점이 별로 없고, 오류 처리 방법을 알고 있는 더 높은 층위의 코드로 오류를 전송해야 하는 경우가 많음

### 복구할 수 없는 오류

- 개발자가 코드의 어느 부분에서 '뭔가를 망쳐놓은 것'
  - 코드와 함께 추가되어야 하는 리소스가 없음
  - 잘못된 입력 인수로 호출
  - 일부 필요한 상태를 사전에 초기화하지 않음

### 호출하는 쪽에서만 오류 복구 가능 여부를 알 때가 많다

- 대부분의 오류는 한 코드가 다른 코드를 호출할 때 발생

```java
class PhoneNumber {
    ...
    static PhoneNumber parse(String number){
        if(!isValidPhoneNumber(number)) {
            ...
        }
    }
}

PhoneNumber getHeadOfficeNumber() {
    return PhoneNumber.parse("0-123asd");
}

PhoneNumber getUserPhoneNumber(UserInput input) {
    return PhoneNumber.parse(input.getPhoneNumber());
}
```

- 함수를 호출할 때 잘못된 전화번호를 하드 코드로 넣어놨다면 이는 프로그래밍 오류이며 복구할 수 없는 오류이다
- 사용자가 번호를 입력하고 이 값을 인수로 이 함수를 호출한다면 입력이 잘못된 전화번호는 프로그램이 복구할 수 있고, 복구해야하는 오류
  - 오류메시지를 UI에 표시하는 것이 최상
- 유효하지 않은 전화번호에 대해 프로그램이 복구할 수 있는지 여부는 오직 PhoneNumber.parse() 함수를 호출하는 코드만이 알 수 있음
- 다음 중 하나라도 해당되는 경우 함수에 제공된 값으로 인해 발생하는 오류는 호출하는 쪽에서 복구하고자 하는 것으로 간주해야 함
  - 함수가 어디서 호출될지 그리고 호출 시 제공되는 값이 어디서 올지 정확한 지식이 없음
  - 코드가 미래에 재사용될 가능성이 희박함

### 견고성 vs 실패

- 오류가 발생할 때 다음 중 하나를 선택해야 함
  - 실패. 더 높은 코드 계층이 오류를 처리하게 하거나 전체 프로그램의 작동을 멈추게 함
  - 오류를 처리하고 계속 진행

### 신속하게 실패하라

- 코드가 어디에서 잘못된 동작을 하는지 파악하거나 스택 트레이스에서 코드 줄 번호를 보면 어디에서 코드가 망가지기 시작하는지 알 수 있다
- 신속하게 실패하기(failing fast)는 가능한 한 문제의 실제 발생 지점으로부터 가까운 곳에서 오류를 나타내는 것
  - 복구할 수 있는 오류의 경우 호출하는 쪽에서 오류로부터 훌륭하고 안전하게 복구할 수 있는 기회를 최대한으로 제공
  - 복구할 수 없는 오류의 경우 개발자가 문제를 신속하게 파악하고 해결할 수 있는 기회를 최대한 제공
- 잘못된 인수로 함수를 호출하는 경우 실패의 신속한 표시는 그 함수가 잘못된 입력과 함께 호출되는 즉시 오류를 발생시키는 것을 의미
- 오류가 실제 바랭한 위치로부터 멀리 떨어진 곳에서 나타나면 오류를 발견하기 위해 역방향을 코드를 찾는 노력을 상당히 많이 기울여야 함
- 실패가 신속하게 이뤄지면, 오류는 실제 위치 근처에서 나타나며 스택 트레이스틑 종종 해당 오류의 위치에 대한 정확한 코드 위치를 제공함

### 요란하게 실패하라

- 오류가 발생하는데도 불구하고 아무도 모르는 상황을 막아야 한다

### 오류를 숨기지 않음

- 어떤 때는 실수를 숨기고 아무 일도 없엇던 것처럼 동작하도록 코드를 작성하고 싶은 마음이 생길 수 있음
- 오류를 숨기는 것은 복구할 수 있는 오류와 복구할 수 없는 오류 모두에 문제를 일으킴

#### 기본값 반환

- 오류가 발생하고 함수가 원하는 값을 반환할 수 없는 경우 기본값을 반환하는 것이 간단하고 쉬운 해결책처럼 보일 때가 있음
  - 적절한 에러 전달과 처리를 위해 코드를 추가하는 것은 많은 노력이 드는 것처럼 보임
- 기본값의 문제점은 오류가 발생했다는 사실을 숨기는 것
  - 코드를 호출하는 쪽에서 모든 것이 정상인 것처럼 계속 진행함

```java
class AccountManager {
    private final AccountStore accountStore;
    
    Double getAccountBalancedUsd(Integer customerId) {
        AccountResult result = accountStore.lookup(customerId);
        if(!result.success()) {
            return 0.0;
        }
        return result.getAccount().getBalancedUsd();
    }
}
```

- 실제 잔액이 0원인지 오류가 발생한 것인지 확인할 방법이 없음
- 코드에 기본값을 두는 것이 유용한 경우가 있을 수 있지만, 오류를 처리할 때는 대부분의 경우 적합하지 않음

#### 널 객체 패턴

- 개념적으로 기본값과 유사
- 실제 반환값처럼 보이지만 모든 멤버 함수는 아무것도 하지 않거나 의미 없는 기본값을 반환함

```java
class InvoiceManager {
    private final InvoiceStore invoiceStore;
    
    List<Invoice> getUnpaidInvoices(Integer customerId) {
        InvoiceResult result = invoiceStore.query(customerId);
        if(!result.success()) {
            reutnr new ArrayList<Invoive>;
        }
        return result.getInvoices().filter(invoice -> !invoice.isPaid());
    }
}
```

#### 아무것도 하지 않음

- 코드가 무언가를 반환하지 않고 단지 어떤 작업을 수행하는 경우 문제가 발생할 때 가능한 한 가지 옵션은
오류가 발생했다는 신호를 보내지 않는 것
- 호출하는 쪽에서는 코드에서 작업이 의도대로 완료되었다고 가정하기 때문에 일반적으로 이렇게 하는 것은 바람직하지 않음

```java
class MutableInvoice {
    void addItem(InvoiceItem item) {
        if(item.getPrice().getCurrency() != this.getCurrency()) {
            return;
        }
        this.items.add(item);
    }
}

class InvoiceSender {
    private final EmailService emailService;
    
    void emailInvoice(String emailAddress, Invoice invoice) {
        try {
            emailService.sendPlainText(emailAddress, InvoiceFormat.plainText(invoice));
        }catch (EmailException e) {}
    }
}
```

- 오류를 로그로 작성해도 큰 의미가 없음
  - 개발자만 알고 실 사용자는 정상작동이라 가정할 수 있음


------------

## 오류 전달 방법

- 오류가 발생하면 일반적으로 더 높은 계층으로 오류를 알려야 함
- 오류로부터 복구할 수 없는 경우 프로그램의 훨씬 더 높은 계층에서 실행을 중지, 오류를 기록하거나 전체 프로그램의 실행을 종료하는 것을 의미
- 오류로부터 복구가 잠재적으로 가능한 경우 일반적으로 즉시 호출하는 쪽(또는 호출 체인에서 한두 수준 위의 호출자)에 오류를 알려 정상적으로 처리할 수 있도록 해야 함
- 오류를 알리는 방법 두 가지
  - 명시적 방법
    - 코드를 직접 호출한 쪽에서 오류가 발생할 수 있음을 인지할 수 있게 함
    - 어떻게 처리할지는 호출하는 쪽에 달려 있음
    - 오류가 발생할 가능성이 코드 계약의 명확한 부분에 나타나 있기 때문에 오류를 모르고 넘어갈 수 있는 방법은 거의 없음
  - 암시적 방법
    - 코드를 호출하는 쪽에 알리지만 호출하는 쪽에서 그 오류를 신경 쓰지 않아도 됨
    - 코드 계약의 숨겨진 세부 조항

### 요약: 예외

- 많은 프로그래밍 언어들은 오류나 예외적인 상황이 발생한 경우 이를 전달하기 위한 방법으로 예외라는 개념을 갖고 있음
- 예외가 발생할 때 콜 스택을 거슬러 올라가는데 예외를 처리하는 코드를 만나거나 더 이상 콜 스택이 없을 때까지 그렇게 함
  - 올라갈 콜 스택이 없는 경우에는 오류 메시지를 출력하고 프로그램이 종료됨
- 자바는 검사 예외(checked exception)와 비검사 예외(unchecked exception)의 개념을 모두 가지고 있음

### 명시적 방법: 검사 예외

- 검사 예외를 사용하는 것은 오류를 전달하기 위한 명시적인 방법

#### 검사 예외를 사용한 오류 전달

```java
class NegativeNumberException extends Exception {
    private final Double erroneousNumber;
    
    NegativeNummberException(Double erroneousNumber) {
        this.erroneousNumber = erroneousNumber;
    }
    
    Double getErroneousNumber() {
        return erroneousNumber;
    }
}

Double getSquareRoot(Double value) throws NegativeNumberException {
    if (value < 0.0) {
        throw new NegativeNumberException(value);
    }
    return Math.sqrt(value);
    
}
```
- NegativeNumberException은 오류를 전달할 뿐만 아니라 오류를 초래한 값도 멤버 변수로 가질 수 있는데
이 경우 디버깅에 도움이 됨
- getSquareRoot() 함수의 시그니처에는 throws NegativeNumberException이 포함되어 있고 이것은 이 함수가
검사 예외를 발생시킬수 있다는 것을 의미함

#### 검사 예외 처리

- getSquareRoot() 함수를 호출하는 코드는 NegativeNumberException 예외를 처리하거나 함수 시그니처에 이 예외를 발생시킰 수 있음을
표히새야 함

```java
void displaySquareRoot() {
    Double value = ui.getInputNumber();
    try {
        ui.setOutput("Square root is : " + getSquareRoot(value));
    } catch (NegativeNumberException e) {
        ui.setError("Can't get square root of negative number: " + e.getErroneousNumber());
    }
}
```

- 예외를 포착하지 않는 경우에는 이 함수의 시그니처에 이 예외가 발생할 수 있음을 선언해야 함
  - 이 경우에는 예외가 발생할 경우 처리를 자신이 하지 않고 자기 자신을 호출하는 코드에게 맡기는 것

```java
void displaySquareRoot() throws NegativeNumberException {
    Double value = ui.getInputNumber();
    ui.setOutput("Square root is : " + getSquareRoot(value));
}
```

- displaySquareRoot() 함수가 예외를 포착하지도 않고, 자신의 함수 시그니처에 선언하지도 않으면 컴파일이 되지 않음
- 호출하는 쪽에서 어떤 형태로든 해당 오류를 강제적인 방식으로 인지할 수 밖에 없기 때문에 검사 예외는 오류를 전달하는 명시적 방법이 됨

### 암시적 방법: 비검사 예외

- 비검사 예외를 사용하면 다른 개발자들은 코드가 이 예외를 발생시킬 수 있다는 사실을 전혀 모를 수 있음
- 비검사 예외는 오류가 발생할 수 없다는 것을 호출하는 쪽에서 인지하리라는 보장이 없기 때문에 오류를 암시적으로 알리는 방법

#### 비검사 예외를 사용한 오류 전달

```java
class NegativeNumberException extends RuntimeException {
    private final Double erroneousNumber;

    NegativeNummberException(Double erroneousNumber) {
      this.erroneousNumber = erroneousNumber;
    }
  
    Double getErroneousNumber() {
      return erroneousNumber;
    }
}

/**
 * 
 * @throws NegativeNumberException 값이 음수일 경우
 */

Double getSquareRoot(Double value) {
    if (value < 0.0) {
      throw new NegativeNumberException(value);
    }
    return Math.sqrt(value);
}
```

#### 비검사 예외 처리

- getSquareRoot() 가 NegativeNumberException 예외를 발생시키면 이 예외를 처리하는 코드를 만날 때까지 계속 올라가거나,
끝까지 그 코드가 없으면 프로그램이 종료됨

### 명시적 방법: 널값이 가능한 반환 유형

#### 널값을 이용한 오류 전달

```java
Optional<Double> getSquareRoot(Double value) {
    if(value < 0.0) {
        return Optional.empty();
    }
    reutnr Math.sqrt(value);
}
```
- 널값을 반환할 때의 한 가지 문제점은 오류가 발생한 이유에 대한 정보를 제공하지 않기 때문에 널값이 의미하는 바를
설명하기 위해 주석문이나 문서를 추가해야 함

### 명시적 방법: 리절트 반환 유형

- 널값이나 옵셔널 타입을 반환할 때의 문제 중 하나는 오류 정보를 전달할 수 없다는 것
- 스위프트, 러스트, F# 같은 언어들은 리절트 유형을 지원하고 사용하기 쉽도록 편한 구문을 제공하지만
직접 만들어야 할 수 도 있음

```java
class Result<V, E> {
  private final Optional<V> value;
  private final Optional<E> error;

  private Result(Optional<V> value, Optional<E> error) {
    this.value = value;
    this.error = error;
  }

  static Result<V, E> ofValue(V value) {
    return new Result<V, E>(Optional.of(value), Optional.empty());
  }
  
  static Result<V, E> ofError(E Error) {
      return new Result<V, E>(Optional.empty(), Optional.of(error));
  }
  
  Boolean hasError() {
      return error.isPresent();
  }
  
  V getValue() {
      return value.get();
  }
  
  E getError() {
      return error.get();
  }
}
```

- 이 클래스는 값이나 오류 중 한 가지로만 인스턴스를 만들 수 있고 둘 다 가능한 인스턴스는 생성할 수 없다
- 자신만의 리절트 유형을 정의한다면 제대로 사용될지 여부는 다른 개발자가 리절트 유형을 사용하는 방법에 얼마나 익숙해지는가에 달려 있음

#### 리절트 유형을 이용한 전달

```java
class NegativeNumberError extends Error {
  private final Double erroneousNumber;

  NegativeNummberException(Double erroneousNumber) {
    this.erroneousNumber = erroneousNumber;
  }

  Double getErroneousNumber() {
    return erroneousNumber;
  }
}

Result<Dboule, NegativeNumberError> getSquareRoot(Double value) {
    if (value < 0.0) {
        return Result.ofError(new NegativeNumberError(value));
    }
    return Result.ofValue(Math.sqrt(value));
}
```

#### 리절트 처리

```java
void displaySquareRoot() {
    Result<Double, NegativeNumberError> squareRoot = getSquareRoot(ui.getInputNumber());
    if(squareRoot.hasError()) {
        ui.setError("Can't get square root of a negative number : " + squareRoot.getError().getErroneousNumber());
    }else {
        ui.setOutput("Square root is: " + squareRoot.getValue());
    }
}
```

### 명시적 방법: 아웃컴 반환 유형

- 어떤 함수들은 값을 반환하기보다는 단지 무언가를 수행하고 값을 반환하지 않음
- 어떤 일을 하는 동안 오류가 발생할 수 있고 그것을 호출한 쪽에 알리고자 한다면 함수가 수행한 동작의 결과를 나타내는 값을
반환하도록 함수를 수정하는 것이 한 가지 방법이 될 수 있음
- 아웃컴 반환 유형을 반환할 때 호출하는 쪽에서 반환값을 강제적으로 확인해야 한다면 오류를 알리는 명백한 방법

#### 아웃컴을 이용한 오류 전달

```java
Boolean sendMessage(Channel channel, String message) {
    if(channel.isOpen()) {
        channel.send(message);
        return true;
    }
    return false;
}
```
- 채널이 열려있지 않으면 오류라 오류가 발생했음을 전달함
- 단순한 불리언보다 더 정교한 아웃컴 유형을 사용하는 경우도 있음
- 이럴경우 열거형을 사용하면 유용

#### 아웃컴 처리

```java
void sayHello(Channel channel) {
    if (sendMessage(channel, "hello")) {
        ui.setOutput("Hello sent");
    } else {
        ui.setError("Unable to send hello");
    }
}
```

#### 아웃컴이 무시되지 않도록 보장

- 아웃컴 반환 유형에 대한 문제점 중 하나는 호출하는 쪽에서 반환값을 무시하거나 함수가 값을 반환한다는 사실조차
인식 못할 수 있어 명시적 방법으로서 한계가 있음
- 일부 언어에서는 호출하는 쪽에서 함수의 반환값을 무시하면 컴파일러가 경고를 생성하도록 함수를 표시할 수 있음
  - 자바의 CheckReturnValue 애너테이션
  - C#의 MustUseResultValue 애너테이션
  - C++의 [[nodiscard]] 속성

```java
@CheckReturnValue
Boolean sendMessage(Channel channel, String message){
  ...
}
```

### 암시적 방법: 프로미스 또는 퓨처

- 비동기적으로 실행하는 코드를 작성할 때 프로미스나 퓨처를 반환하는 함수를 작성하는 것이 일반적임
- 프로미스나 퓨처를 사용할 때 오류 처리를 강제로 해야 하는 것은 아니기 때문에 암시적인 전달 방법임

#### 프로미스를 이용한 전달

- 함수 안에서 오류가 발생하면 프로미스는 거부(rejected)됨.

```java
Promise<Double> getSquareRoot(Double value) async {
    
}
```

#### 왜 프로미스는 암묵적인 오류 전달 기법인가

- 오류가 발생하고 프로미스가 거부될 수 있음을 알려면 프로미스를 생성하는 함수의 세부 조항이나 구현 세부 사항을 확인해야 함

#### 프로미스를 명시적으로 만들기

- 리절트 유형의 프로미스를 반환하는 것이 한 가지 방법이 될 수 있음
- 유용한 기술이지만 코드가 복잡해짐

### 암시적 방법: 매직값 반환

- 매직값(magic value)(또는 오류 코드)은 함수의 정상적인 반환 유형에 적합하지만 특별한 의미를 부여하는 값
- 매직값이 반환될 수 있다는 것을 알려면 문서나 코드를 읽어야 하기 때문에 암시적 오류 전달 기법

```java
// 음숫값이 입력으로 제공되면 -1을 반환함
Double getSquareRoot(Double value) {
    if (value < 0.0) {
        return -1.0;
    }
    return Math.sqrt(value);
}
```
- 매직값은 코드 계약의 명백한 부분을 통해 호출하는 쪽에 알릴 수 없어서 예상을 벗어나는 결과를 가져올 수도 있고 버그로 이어질 수 있음
- 오류를 알리는 좋은 방법이 아님

--------------

## 복구할 수 없는 오류의 전달

- 현실적으로 복구할 가능성이 없는 오류가 발생하면 신속하게 실패하고 요란하게 실패하는 것이 최상
- 비검사 예외를 발생
- 프로그램이 패닉(panic)이 되도록(지원하는 언어를 사용하는 경우)
- 체크나 어서션의 사용

------------

## 호출하는 쪽에서 복구하기를 원할 수도 있는 오류의 전달

- 팀이 동의한 철학이 다른 어떤 주장보다도 중요함

### 비검사 예외를 사용해야 한다는 주장

- 잠재적으로 복구할 수 있는 오류에도 비검사 예외를 사용하는 것이 더 나은 이유에 대한 일반적인 주장은 다음과 같음

#### 코드 구조 개선

- 대부분의 오류 처리가 코드의 상위 계층에서 이루어질 수 있기 때문에 비검사 예외를 발생시키면 코드 구조를 개선할 수 있다고 주장
- 오류가 높은 계층까지 거슬러 올라오면서 전달되고 그 사이에 있는 코드는 오류 처리를 할 필요가 없음
- 사용자 응용 프로그램이라면 오류 처리 계층은 오류 메시지를 UI에 표시할 수 있고, 서버나 백엔드 프로세서라면 오류 메시지가 기록될 수 있음
- 핵심 장점은 오류를 처리하는 로직이 코드 전체에 퍼지지 않고 별도로 몇 개의 계층에만 있다는 점

#### 개발자들이 무엇을 할 것인지에 대해서 실용적이어야 함

- 일부에서는 개발자들이 너무 많은 명시적 오류 전달을 접근하면 결국 잘못된 일을 한다고 주장
  - 예외를 포착하고도 무시하거나 널이 가능한 유형을 확인도 하지 않고 널이 불가능한 유형으로 변환을 하는 것
- 변화하는 과정에서 예외를 추가하게 된다면 시그니처에 추가해야하고, 이 함수를 호출하는 모든 코드를 수정하고 그보다 더 위 계층의 코드까지도 수정해야 할 수도 있음
  - 작업의 양이 너무 많아짐
  - 따라서 오류를 숨길 가능성이 커짐
- 명시적인 오류 전달 방식을 사용하면 코드의 계층을 올라가면서 오류를 반복적으로 전달하고 이를 처리하는 일련의 작업이 필요한데,
이런 번거로운 작업 대신 개발자는 편의를 도모하고 잘못된 작업을 하고 싶은 마음이 들 수 있음
- 비검사 예외의 사용에 찬성하기 위해 자주 표현되는 주장 중 하나가 이 문제에 대해 실용적으로 접근해야 한다는 주장

### 명시적 기법을 사용해야 한다는 주장

#### 매끄러운 오류 처리

- 비검사 예외를 사용한다면 모든 오류를 매끄럽게 처리할 수 있는 단일 계층을 갖기가 어려움
  - 사용자 입력이 잘못되면 해당 입력 필드 바로 옆에 오류메시지를 보여주는 것이 타당
  - 입력을 처리하는 코드를 작성하는 엔지니어가 오류 시나리오를 알지 못하고 더 높은 수준으로 전달되도록 내버려 둔다면
  사용자 친화적이지 않은 오류 메시지를 UI에 표시할 수 있음
- 호출하는 쪽에 잠재적 오류를 강제적으로 인식하도록 하면 이러한 오류를 좀 더 매끄럽게 처리할 가능성이 커짐

#### 실수로 오류를 무시할 수 없다

- 실제로 오류를 처리해야 하는 경우 비검사 예외가 사용되면 오류를 처리하지 않는 일이 일어나기 쉬움

### 필자의 의견: 명시적 방식을 사용하라

- 호출하는 쪽에서 복구하기를 원할 수도 있는 오류에 대해 비검사 예외를 사용하지 않는 것이 최상
- 비검사 예외의 사용은 코드베이스 전반에 걸쳐 완전히 문서화되는 경우가 매우 드물며, 의미하는 바는 해당 함수에 대해 발생 가능한 오류와
이에 대한 처리를 어떻게 해야 하는지 개발자가 확실하게 알기란 거의 불가능

----------

## 컴파일러 경고를 무시하지 말라

- 컴파일러는 오류뿐만 아니라 대부분의 컴파일러는 경고 메시지도 출력함
- 버그에 대한 조기 경고일 수 있음
- 경고가 실제로 걱정할 것이 아닌 경우에는 일반적으로 특정 경고만 억제할 수 있는 방법이 있음

```java
class UserInfo {
    private final String realName;
    
    // 실제 이름을 사용하지 않도록 마이그레이션 작업중이라 displayName이 지금은 사용되지 않음
    // 곧 사용될 것을 대비해 만들어놓은 변수
    @SuppressWarnings("unused")
    private final String displayName;
}
```

----------

## 요약

- 오류에는 크게 두 가지 종류가 있다
  - 시스템이 복구할 수 있는 오류
  - 시스템이 복구할 수 없는 오류
- 해당 코드에 의해 생성된 오류로부터 복구할 수 있는지 여부를 해당 코드를 호출하는 쪽에서만 알 수 있는 경우가 많다
- 에러가 발생하면 신속하게 실패하는 것이 좋고, 에러를 복구할 수 없는 경우에는 요란하게 실패하는 것이 바람직하다
- 오류를 숨기는 것은 바람직하지 않을 때가 많으며, 오류가 발생했다는 신호를 보내는 것이 바람직하다
- 오류 전달 기법은 두 가지 범주로 나눌 수 있다
  - 명시적 방법: 코드 계약의 명확한 부분, 호출하는 쪽에서는 오류가 발생할 수 있음을 인지한다.
  - 암시적 방법: 코드 계약의 세부 조항을 통해 오류에 대한 설명이 제공되거나 전혀 설명이 없을 수도 있다.
- 복구할 수 없는 오류에 대해서는 암시적 오류 전달 기법을 사용해야 한다
- 잠재적으로 복구할 수 있는 오류에 대해서는
  - 명시적 혹은 암시적 기법 중 어느 것을 사용할지에 대해서는 개발자들 사이에서도 일치되는 의견이 없다
  - 필자의 의견으로는 명시적인 기법이 사용되어야 한다
- 컴파일러 경고는 종종 코드에 문제가 있을 때 이에 대해 표시해준다. 경고에 주의를 기울이는 것이 바람직하다.